src/app/zap-templates/templates/chip/python-ChipDeviceController-ClusterCommands.zapt

```diff
@@ -34,18 +35,26 @@ extern "C" {
 {{#chip_server_clusters}}
 // Cluster {{asCamelCased name false}}
 {{#chip_server_cluster_commands}}
-CHIP_ERROR chip_ime_AppendCommand_{{asCamelCased clusterName false}}_{{asCamelCased name false}}(chip::DeviceController::ChipDeviceController * dc,chip::EndpointId ZCLendpointId, chip::GroupId ZCLgroupId{{#chip_server_cluster_command_arguments}}, {{chipType}} {{asCamelCased label}}{{#if (isByteString type)}}, uint32_t {{asCamelCased label}}_Len{{/if}}{{/chip_server_cluster_command_arguments}})
+CHIP_ERROR chip_ime_AppendCommand_{{asCamelCased clusterName false}}_{{asCamelCased name false}}(chip::Controller::Device * device, chip::EndpointId ZCLendpointId, chip::GroupId ZCLgroupId{{#chip_server_cluster_command_arguments}}, {{chipType}} {{asCamelCased label}}{{#if (isByteString type)}}, uint32_t {{asCamelCased label}}_Len{{/if}}{{/chip_server_cluster_command_arguments}})
 {
-    return app::cluster::{{asCamelCased clusterName false}}::Encode{{asType name}}Command(dc->GetCommandSender(), ZCLendpointId, ZCLgroupId{{#chip_server_cluster_command_arguments}}, {{asCamelCased label}}{{#if (isByteString type)}}, {{asCamelCased label}}_Len{{/if}}{{/chip_server_cluster_command_arguments}});
+    if (device == nullptr)
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    return app::cluster::{{asCamelCased clusterName false}}::Encode{{asType name}}Command(device->GetCommandSender(), ZCLendpointId, ZCLgroupId{{#chip_server_cluster_command_arguments}}, {{asCamelCased label}}{{#if (isByteString type)}}, {{asCamelCased label}}_Len{{/if}}{{/chip_server_cluster_command_arguments}});
```

Comment:
I think what I was saying is to have something that leverage the current code in the tree.

For example currently the controller API looks like:
```
// A cluster instance is associated with a particular device instance
chip::Controller::OnOffCluster cluster;
cluster.Associate(device, endPointId);

...

// Then a command is encoded and sent
cluster.On(nullptr /* completion callback */);
```

I guess that in your case, it could be pretty similar:
```
// A cluster instance is associated with a particular device instance
chip::Controller::OnOffCluster onOffCluster;
chip::Controller::BarrierControl barrierControlCluster;
onOffCluster.Associate(device, endPointId);
barrierControlCluster.Associate(device, endPointId);

...

// Then a bunch of commands are added
onOffCluster.On(nullptr /* completion callback */);
barrierControlCluster.BarrierControlStop(nullptr /* completion callback */);
...

// The commands can then be sent
device->SendCommands();
```

Inside `ChipCluster` methods there could be an `#ifdef CHIP_APP_USE_INTERACTION_MODEL` that does different things depending if IM is turned on/off.

And in order to make it easier to land this changes and get it to work with all controller, I would first focus on getting one command encoded with the IM encoding scheme. That should let you keep the exact same version of the embedder code for the controller for the IM and non-IM version.

Saying that more clearly, inside `ChipClusters.cpp` it could be generated as:
```
CHIP_ERROR OnOffCluster::Off(Callack::Callback<> * onCompletion)
{
#ifdef CHIP_APP_USE_INTERACTION_MODEL
    // AppendCommands and SendCommands are both methods of ChipDevice that call the necessary pieces of CommandSender
    AppendCommand(mEndpoint, onCompletion);
    return SendCommands();
#else
    SystemPacketBufferHandle payload = encodeOnOffClusterOnCommand(mEndpoint);
    return SendCommand(std::move(payload), onCompletion);
#endif
}
```

From here it could be possible to focus on how to get the receive message down to ZCL.

Reply:
Done